src/serialize.rs

```diff
@@ -0,0 +1,397 @@
+///     ) -> Result<(), S::Error>
+///     where
+///         S: serde::ser::SerializeMap,
+///     {
+///         // Call `try_serialize` for every serializable component we want to save
+///         try_serialize::<Position, _, _>(&entity, &ComponentId::Position, map)?;
+///         try_serialize::<Velocity, _, _>(&entity, &ComponentId::Velocity, map)?;
+///         // Or do something custom for more complex cases.
+///         Ok(())
+///     }
+/// }
+/// ```
+pub trait SerializeContext {
+    /// Serialize a single entity into a map
+    fn serialize_entity<S>(&mut self, entity: EntityRef<'_>, map: &mut S) -> Result<(), S::Error>
+    where
+        S: SerializeMap;
+}
+
+/// If `entity` has component `T`, serialize it under `key` in `map`
+///
+/// Convenience method for [`SerializeContext`] implementations.
+pub fn try_serialize<T: Component + Serialize, K: Serialize + ?Sized, S: SerializeMap>(
+    entity: &EntityRef<'_>,
+    key: &K,
+    map: &mut S,
+) -> Result<(), S::Error> {
+    if let Some(x) = entity.get::<T>() {
+        map.serialize_key(key)?;
+        map.serialize_value(&*x)?;
+    }
+    Ok(())
+}
+
+/// Serialize a [`World`] through a [`SerializeContext`] to a [`Serializer`]
+pub fn serialize<C, S>(world: &World, context: &mut C, serializer: S) -> Result<S::Ok, S::Error>
+where
+    C: SerializeContext,
+    S: Serializer,
+{
+    let mut seq = serializer.serialize_map(None)?;
+    for (id, components) in world {
+        seq.serialize_key(&id)?;
+        seq.serialize_value(&SerializeComponents(RefCell::new((
+            context,
+            Some(components),
+        ))))?;
+    }
+    seq.end()
+}
+
+struct SerializeComponents<'a, C>(RefCell<(&'a mut C, Option<EntityRef<'a>>)>);
+
+impl<'a, C: SerializeContext> Serialize for SerializeComponents<'a, C> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut this = self.0.borrow_mut();
+        let entity = this.1.take().unwrap();
+        let mut map = serializer.serialize_map(None)?;
+        this.0.serialize_entity(entity, &mut map)?;
+        map.end()
+    }
+}
+
+/// Deserialize a [`World`] with a [`DeserializeContext`] and a [`Deserializer`]
+pub fn deserialize<'de, C, D>(context: &mut C, deserializer: D) -> Result<World, D::Error>
+where
+    C: DeserializeContext,
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_map(WorldVisitor(context))
+}
+
+/// Implements deserialization of entities from a serde [`MapAccess`] into an [`EntityBuilder`]
+///
+/// Data external to the [`World`] can be populated during deserialization by storing mutable
+/// references inside the struct implementing this trait.
+///
+/// # Example
+/// ```
+/// # use serde::{Serialize, Deserialize};
+/// # #[derive(Deserialize)]
+/// # struct Position([f32; 3]);
+/// # #[derive(Deserialize)]
+/// # struct Velocity([f32; 3]);
+/// use hecs::{*, serialize::*};
+///
+/// struct Context;
```

Comment:
`Context` here does nothing, which makes it a little harder to grok why this trait is the way it is.